### DISCUSSION TITLE
How to use specific version of system-installed Python?

### DISCUSSION THREAD

**Discussion by DUOLabs333** [pyenv/pyenv#2970] [Q&A] [answered]
I know that "system" can be used to tell pyenv to load what ever is the next "python" binary in the PATH. However, is there a way to use the next "python.X.Y" in the PATH? I couldn't see anything in the README about it.

**Comment by native-api** [comment]
"`system`" uses whatever is further on PATH, regardless of the executable name.

**Reply by DUOLabs333** [reply]
So, it looks for the first executable that matches `python*(.|)*`?

**Reply by native-api** [reply, marked as answer]
Pyenv intercepts executables by creating shims (see the README). `pyenv rehash` creates shims for all executables that are installed by all Pyenv-managed versions -- which can be anything (`python*`, `pypy*` for PyPy, `conda*` and some other utilities for Anaconda/Miniconda (with some exceptions for Anaconda); and also any executables created by installed Pip packages).

Whenever you hit a shim when running a command, the shim's logic runs `exec pyenv exec` which ultimately calls `pyenv which <shim's name>` to determine what real executable to run. If the selected version is `system`, `pyenv which` does that by removing shims from PATH and running `command -v <shim's name>` (`command -v` is Bash's built-in equivalent of `which`).

**Comment by DUOLabs333** [comment]
Apparently, the README addresses this concern:
> Note that Pyenv considers those installations outside its control and does not attempt to inspect or distinguish them in any way. So e.g. if you are on MacOS and have OS-bundled Python 3.8.9 and Homebrew-installed Python 3.9.12 and 3.10.2 -- for Pyenv, this is still a single "system" version, and whichever of those is first on PATH under the executable name you specified will be run.

So, as far as I can tell, pyenv does not have the ability to distinguish between multiple installed `system` versions.

**Reply by native-api** [reply]
Since I wrote the previous answer, I've discovered that Homebrew got a subcommand `brew pyenv-sync` -- which creates Pyenv entries for Homebrew-installed Pythons. Its logic seems dubious however: it creates superfluous entries (e.g. for `3.10.2`, it would also create entries "`3.10.1`" and "`3.10.0`"), and I didn't find a built-in way to customize that or remove them other than by uninstalling the corresponding Homebrew-provided Python.
